AX: Support the AXRangesForSearchPredicate API that VoiceOver uses to retrieve misspellings.

AndresGonzalezApple opened the pull request:
#### b590c842a3e221a9b0e9cb154c95676eb09b0da2
<pre>
AX: Support the AXRangesForSearchPredicate API that VoiceOver uses to retrieve misspellings.
<a href="https://bugs.webkit.org/show_bug.cgi?id=278254">https://bugs.webkit.org/show_bug.cgi?id=278254</a>
&lt;<a href="https://rdar.apple.com/problem/134083674">rdar://problem/134083674</a>&gt;

Reviewed by Tyler Wilcock.

Follow up to <a href="https://github.com/WebKit/WebKit/pull/31136.">https://github.com/WebKit/WebKit/pull/31136.</a>

This functionality was originally implemented as an extension of the AXUIElementsForSearchPredicate API, but that didn&apos;t work for VoiceOver since it changes the type of the return value which is unexpected for the middleware system framework. Therfore, a new AXRangesForSearchPredicate API was added to expose the functionality on the Mac.

The corresponding AccessibilityUIElement::textMarkerRangeForSearchPredicate test method was updated.

* Source/WebCore/accessibility/AXSearchManager.cpp:
(WebCore::AXSearchManager::findMatchingRange):
* Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperMac.mm:
(-[WebAccessibilityObjectWrapper accessibilityAttributeValue:forParameter:]):
* Tools/WebKitTestRunner/InjectedBundle/mac/AccessibilityUIElementMac.mm:
(WTR::AccessibilityUIElement::textMarkerRangeForSearchPredicate):

Canonical link: <a href="https://commits.webkit.org/282401@main">https://commits.webkit.org/282401@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9a4bbd482feeadad216f793fdf26580afe7960b4

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/62966 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/42322 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/15562 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/66987 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/13570 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/65086 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/50009 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/13854 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/50758 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/9364 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/66035 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/39323 "Passed tests") | [❌ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/54534 "Found 1 new API test failure: TestWebKitAPI.WKWebExtensionAPIRuntime.ConnectNativeWithInvalidMessage (failure)") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/31440 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/36013 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/11867 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/12446 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/57557 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/12197 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/68682 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/6912 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/11824 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/58074 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/6944 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/54594 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/58281 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/13985 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/5766 "Passed tests") | | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/38142 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/39222 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/40333 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/38964 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

minorninth commented on Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperMac.mm:
```diff
@@ -3248,7 +3248,7 @@ - (id)accessibilityAttributeValue:(NSString*)attribute forParameter:(id)paramete
     }
 
-    if ([attribute isEqualToString:NSAccessibilityUIElementsForSearchPredicateParameterizedAttribute]) {
+    if ([attribute isEqualToString:@"AXRangesForSearchPredicate"]) {
```
Maybe we should define a constant for this? I know it's not done consistently now but I think it'd be a nice goal to get rid of magic strings and constants from the source.

minorninth reviewed (approved):
looks good

(Not a WebKit reviewer)

AndresGonzalezApple committed:
AX: Support the AXRangesForSearchPredicate API that VoiceOver uses to retrieve misspellings.
https://bugs.webkit.org/show_bug.cgi?id=278254
<rdar://problem/134083674>

Reviewed by Tyler Wilcock.

Follow up to https://github.com/WebKit/WebKit/pull/31136.

This functionality was originally implemented as an extension of the AXUIElementsForSearchPredicate API, but that didn't work for VoiceOver since it changes the type of the return value which is unexpected for the middleware system framework. Therfore, a new AXRangesForSearchPredicate API was added to expose the functionality on the Mac.

The corresponding AccessibilityUIElement::textMarkerRangeForSearchPredicate test method was updated.

* Source/WebCore/accessibility/AXSearchManager.cpp:
(WebCore::AXSearchManager::findMatchingRange):
* Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperMac.mm:
(-[WebAccessibilityObjectWrapper accessibilityAttributeValue:forParameter:]):
* Tools/WebKitTestRunner/InjectedBundle/mac/AccessibilityUIElementMac.mm:
(WTR::AccessibilityUIElement::textMarkerRangeForSearchPredicate):

Canonical link: https://commits.webkit.org/282401@main
